fdbserver/DataDistribution.actor.cpp

```diff
@@ -395,18 +425,164 @@ struct DataDistributor : NonCopyable, ReferenceCounted<DataDistributor> {
+				          return a.id < b.id; // Inplacement sort in ascending order
+			          });
+			// Cleanup audit metadata for Failed/Complete audits
+			// Resume RUNNING audits
+			// Keep Error audits persistent
+			for (const auto& auditState : restoredAudits[auditType]) {
+				TraceEvent(SevVerbose, "AuditStorageResumeCheck", self->ddId)
+				    .detail("AuditID", auditState.id)
+				    .detail("AuditType", auditState.getType());
+				if (auditState.getPhase() == AuditPhase::Error) {
+					continue;
+				} else if (auditState.getPhase() == AuditPhase::Failed) {
+					if (numFinishAuditsCleared < numFinishAuditsToClear) {
+						// Clear both audit metadata and corresponding progress metadata
+						self->addActor.send(clearAuditMetadata(self->txnProcessor->context(),
+						                                       auditState.getType(),
+						                                       auditState.id,
+						                                       /*clearProgressMetadata=*/true));
+						numFinishAuditsCleared++;
+					}
+					continue;
+				} else if (auditState.getPhase() == AuditPhase::Complete) {
+					if (numFinishAuditsCleared < numFinishAuditsToClear) {
+						// Clear audit metadata only
+						// No need to clear the corresponding progress metadata
+						// since it has been cleared for Complete audits
+						self->addActor.send(clearAuditMetadata(self->txnProcessor->context(),
+						                                       auditState.getType(),
+						                                       auditState.id,
+						                                       /*clearProgressMetadata=*/false));
+						numFinishAuditsCleared++;
+					}
+					continue;
+				}
+				ASSERT(auditState.getPhase() == AuditPhase::Running);
+				TraceEvent(SevDebug, "AuditStorageResume", self->ddId)
+				    .detail("AuditID", auditState.id)
+				    .detail("AuditType", auditState.getType())
+				    .detail("NumFinishAuditsCleared", numFinishAuditsCleared)
+				    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+				runAuditStorage(self, auditState, 0, "ResumeAudit");
+			}
+		}
+
+		self->auditInitialized.send(Void());
+		TraceEvent(SevDebug, "AuditStorageResumeDone", self->ddId);
+		return Void();
+	}
+
 	// Initialize the required internal states of DataDistributor from system metadata. It's necessary before
 	// DataDistributor start working. Doesn't include initialization of optional components, like TenantCache, DDQueue,
 	// Tracker, TeamCollection. The components should call its own ::init methods.
 	ACTOR static Future<Void> init(Reference<DataDistributor> self) {
+		state bool auditStorageResumed = false;
 		loop {
-			wait(self->waitDataDistributorEnabled());
-			TraceEvent("DataDistributionEnabled").log();
-
 			TraceEvent("DDInitTakingMoveKeysLock", self->ddId).log();
 			wait(self->takeMoveKeysLock());
 			TraceEvent("DDInitTookMoveKeysLock", self->ddId).log();
 
+			if (!auditStorageResumed) {
+				wait(self->resumeStorageAudits(self));
```

Comment:
Do we need to wait here? can we make it asynchronous?

Reply:
No. We have to update audit metadata and clean up audit metadata inside the function.